nameresolution/consul/consul.go

```diff
@@ -53,28 +52,138 @@ type healthInterface interface {
 }
 
 type resolver struct {
-	config resolverConfig
-	logger logger.Logger
-	client clientInterface
+	config   resolverConfig
+	logger   logger.Logger
+	client   clientInterface
+	registry registryInterface
+}
+
+type registryInterface interface {
+	get(service string) *registryEntry
+	expire(service string) // clears slice of instances
+	remove(service string) // removes entry from registry
+	addOrUpdate(service string, services []*consul.ServiceEntry)
+}
+
+type registry struct {
+	entries map[string]*registryEntry
+	mu      sync.RWMutex
+}
+
+type registryEntry struct {
+	services []*consul.ServiceEntry
+	mu       sync.RWMutex
+}
+
+func (r *registry) get(service string) *registryEntry {
+	return r.entries[service]
```

Comment:
You need a RLock() here, as it's not safe to read and write concurrently.

Reply:
good catch, I had a second look and did a general refactor in this area